RecoLocalTracker/SiPixelRecHits/interface/PixelCPEFastParams.h

```diff
@@ -0,0 +1,181 @@
+
+  template <uint32_t N>
+  struct ClusParamsT {
+    uint32_t minRow[N];
+    uint32_t maxRow[N];
+    uint32_t minCol[N];
+    uint32_t maxCol[N];
+
+    int32_t q_f_X[N];
+    int32_t q_l_X[N];
+    int32_t q_f_Y[N];
+    int32_t q_l_Y[N];
+
+    int32_t charge[N];
+
+    float xpos[N];
+    float ypos[N];
+
+    float xerr[N];
+    float yerr[N];
+
+    int16_t xsize[N];  // (*8) clipped at 127 if negative is edge....
+    int16_t ysize[N];
+
+    Status status[N];
+  };
+
+  // all modules are identical!
+  struct CommonParams {
+    float theThicknessB;
+    float theThicknessE;
+    float thePitchX;
+    float thePitchY;
+
+    uint16_t maxModuleStride;
+    uint8_t numberOfLaddersInBarrel;
+  };
+
+  struct DetParams {
+    bool isBarrel;
+    bool isPosZ;
+    uint16_t layer;
+    uint16_t index;
+    uint32_t rawId;
+
+    float shiftX;
+    float shiftY;
+    float chargeWidthX;
+    float chargeWidthY;
+    uint16_t pixmx;  // max pix charge
+
+    uint16_t nRowsRoc;  //we don't need 2^16 columns, is worth to use 15 + 1 for sign
+    uint16_t nColsRoc;
+    uint16_t nRows;
+    uint16_t nCols;
+
+    uint32_t numPixsInModule;
+
+    float x0, y0, z0;  // the vertex in the local coord of the detector
+
+    float apeXX, apeYY;  // ape^2
+    uint8_t sx2, sy1, sy2;
+    uint8_t sigmax[CPEFastParametrisation::kNumErrorBins], sigmax1[CPEFastParametrisation::kNumErrorBins],
+        sigmay[CPEFastParametrisation::kNumErrorBins];  // in micron
+    float xfact[CPEFastParametrisation::kGenErrorQBins], yfact[CPEFastParametrisation::kGenErrorQBins];
+    int minCh[CPEFastParametrisation::kGenErrorQBins];
+
+    Frame frame;
+  };
+
+  template <typename TrackerTopology>
+  struct LayerGeometryT {
+    uint32_t layerStart[TrackerTopology::numberOfLayers + 1];
+    uint8_t layer[pixelTopology::layerIndexSize<TrackerTopology>];
+    uint16_t maxModuleStride;
+  };
+
+  // ES Product to be used by RecHit generation
+  template <typename TrackerTopology>
+  struct ParamsOnDeviceT {
+    using LayerGeometry = LayerGeometryT<TrackerTopology>;
+    using AverageGeometry = pixelTopology::AverageGeometryT<TrackerTopology>;
+
+    CommonParams m_commonParams;
+    // Will contain an array of DetParams instances
+    DetParams m_detParams[1440];
+    LayerGeometry m_layerGeometry;
+    AverageGeometry m_averageGeometry;
+
+    constexpr CommonParams const& __restrict__ commonParams() const { return m_commonParams; }
+    constexpr DetParams const& __restrict__ detParams(int i) const { return m_detParams[i]; }
+    constexpr LayerGeometry const& __restrict__ layerGeometry() const { return m_layerGeometry; }
+    constexpr AverageGeometry const& __restrict__ averageGeometry() const { return m_averageGeometry; }
+
+    ALPAKA_FN_ACC uint8_t layer(uint16_t id) const {
+      return m_layerGeometry.layer[id / TrackerTopology::maxModuleStride];
+    };
+  };
+
+  template <typename TDev, typename TrackerTraits>
+  class PixelCPEFastParams {
```

Comment:
This class seems to be only a wrapper for a buffer, and the actual data product of interest is the object held by the buffer. Would it be useful if the EventSetup system would support a case where the ESProducer can produce the host-side data product as alpaka host buffer (instead of `std::unique_ptr` etc), the system would implicitly copy it to am alpaka device buffer, and the consuming module would consume the consume the product directly (i.e. in this case the consumer would use a token of type `device::ESGetToken<pixelCPEforDevice::ParamsOnDeviceT<TrackerTraits>, TkPixelCPERecord>`)?

(I'm thinking longer term evolution beyond this CUDA->Alpaka migration, i.e. I would not change this part of this PR)